### PR TITLE
fix(rust): handle error returned by node

### LINKED ISSUE #4662: `#[ockam::node]` macro doesn't handle returned errors
### Observed behavior

When `main` function returns an Error, nothing happens. Node is not stopped, nothing is printed to the log.

### Steps to reproduce

```rs
use ockam::{Context, Error, Result};
use ockam::errcode::{Kind, Origin};

#[ockam::node]
async fn main(mut ctx: Context) -> Result<()> {
    return Err(Error::new_without_cause(Origin::Core, Kind::Unknown));
}
```

### Expected behavior

Error is reported to the log. Node is shutdown automatically

### Ockam Version

0.83.0

### PR DESCRIPTION
**Pull request by adiSuper94**
fixes #4662 

## Current behavior
Error thrown by node with `#[ockam#node]` macro is not handled

## Proposed changes
Attempt to handle it.
To do that we are having to add some constraint on what an `#[ockam#node]` function is allowed to return.
We are restricting it to return `Result<T, Error>` 

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] There are no Merge commits in this Pull Request. Ockam repo maintains a linear commit history. We merge Pull Requests by rebasing them onto the develop branch. Rebasing to the latest develop branch and force pushing to your Pull Request branch is okay.
- [x] I have read and accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have read and accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/.github/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam/blob/develop/.github/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam](https://github.com/build-trust/ockam) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam/edit/develop/.github/CONTRIBUTORS.csv) file in the github web UI and create a separate Pull Request, this will mark the commit as verified.

### CODE REVIEW

**Comment by adiSuper94** [comment]
@SanjoDeundiak how does this look for a starting point?

I also need to check if I can use send some other message that will shut the router down.

I will make changes to the examples if this approach is ok.

File: implementations/rust/ockam/ockam_node/src/executor.rs [hunk 1]
```diff
@@ -113,6 +115,21 @@ impl Executor {
...
+    async fn wrapper<F, T>(sender: SmallSender<NodeMessage>, future: F) -> Result<T>
+    where
+        F: Future<Output = Result<T>> + Send + 'static,
+        // T: Sync + Send + 'static,
+    {
+        match future.await {
+            Ok(val) => Ok(val),
+            Err(e) => {
+                // let sender = self.sender();
+                let _ = sender.send(NodeMessage::AbortNode).await;
```
Comment (SanjoDeundiak) [inline implementations/rust/ockam/ockam_node/src/executor.rs]: It looks like `AbortNode` is not handled properly and doesn't really stop a node (should be a bug, will create an issue for that).
However, you can find the correct way to stop a node inside `Context::stop` function.

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]
I tried it, and it looks like returning an error from a main function doesn't really stop a node. I left a corresponding comment. Also, could you please run `cargo clippy` and take a look if examples compilation can be easily fixed?

**Comment by adiSuper94** [comment]
> However, you can find the correct way to stop a node inside Context::stop function

I tried this, it works!! 

> It looks like AbortNode is not handled properly and doesn't really stop a node (should be a bug, will create an issue for that).

That would be great.

> could you please run cargo clippy and take a look if examples compilation can be easily fixed?

I tried doing that, did not show any errors. But errors do exist. clippy doesn't seem to find them. basically returning `anyhow::Result<T>` is not allowed anymore. I have fixed one such example test locally. ~~Will do the rest too.~~ 
Looks like there is only one example where node returns `anyhow::Result`

**Review by SanjoDeundiak — DISMISSED** [review]

**Comment by SanjoDeundiak** [comment]
@adiSuper94 I added small improvements and squashed the commits. We can merge as soon as CI checks succeed

**Comment by adiSuper94** [comment]
@SanjoDeundiak , thanks for pushing this.

I have a question though, I ran 'cargo clippy' from ockam root. And I did not get any failures. 
So what is the way to find these errors in local. 
Cuz running cargo test, makes my laptop sound like a helicopter 😅

**Comment by SanjoDeundiak** [comment]
Created an issue for `AbortNode` handling https://github.com/build-trust/ockam/issues/6615

**Comment by SanjoDeundiak** [comment]
@adiSuper94 that's weird. For me it showed the problems. Do you think you may have an outdated version?

**Comment by adiSuper94** [comment]
Do you mean outdated version of code or `clippy` ??
I am not sure. I know that I have the latest rust toolchain installed. But ockam project sets the required version to `1.71` i think.

**Review by etorreborre — APPROVED** [review]

**Comment by SanjoDeundiak** [comment]
@adiSuper94 you can also try `clippy --no-deps --all-targets`

**Comment by adiSuper94** [comment]
I haven't pulled the changes you made to my local yet. So I'll try this and see if I get those errors  you fixed. Thanks!!

**Comment by adiSuper94** [comment]
@SanjoDeundiak `cargo clippy` did find the errors when I ran it with the flags you suggested.

**Pull request merged**